map-storage/src/Upload/S3FileSystem.ts

```diff
@@ -58,6 +61,70 @@ export class S3FileSystem implements FileSystemInterface {
         } while (listObjectsResponse.IsTruncated);
     }
 
+    async exist(virtualPath: string): Promise<boolean> {
+        return (await this.listFiles(virtualPath)).length > 0;
+    }
+
+    async move(virtualPath: string, newVirtualPath: string): Promise<void> {
+        return await this.moveInternal(virtualPath, newVirtualPath, true);
+    }
+
+    async copy(virtualPath: string, newVirtualPath: string): Promise<void> {
+        return await this.moveInternal(virtualPath, newVirtualPath);
+    }
+
+    async moveInternal(virtualPath: string, newVirtualPath: string, deletePrevious = false) {
+        let isTruncated = true;
+        let continuationToken: string | undefined = undefined;
+        let result;
+        while (isTruncated) {
+            result = await this.s3.send(
+                new ListObjectsV2Command({
+                    Bucket: this.bucketName,
+                    Prefix: virtualPath,
+                    ContinuationToken: continuationToken,
+                })
+            );
+
+            // Get the list of objects from the result
+            const objects = result.Contents;
+
+            // Use Promise.all to run the copy and delete operations in parallel
+            if (objects) {
+                await Promise.all(
```

Comment:
Running all queries in parallel might hurt performance. If ListObjectsV2Command returns 1000 files, we will run 1000 Copy commands in parallel. That might be ok, but it might also break.

The best thing to do would be to limit parallelism to somewhere between 10 and 100 concurrent requests.

Hopefully, you can do this easily with the p-limit package that is already used exactly for this in the UploadController.